src/ray/raylet/main.cc

```diff
@@ -159,10 +178,14 @@ int main(int argc, char *argv[]) {
   boost::asio::io_service main_service;
 
   //  initialize mock gcs & object directory
-  auto gcs_client = std::make_shared<ray::gcs::AsyncGcsClient>(redis_address, redis_port,
-                                                               redis_password);
-  RAY_LOG(DEBUG) << "Initializing GCS client "
-                 << gcs_client->client_table().GetLocalClientId();
+  ray::gcs::ClientOption client_option(redis_address, redis_port, redis_password);
+  ray::rpc::ClientTableData raylet_info =
+      std::move(PopulateRayletInfo(object_manager_config, node_manager_config));
```

Comment:
Some states are still unavailable here. E..g, `node_manager_port` is 0 here, and will be set when the node manager grpc server starts (in the constructor of node manger). 
That's why we get these attributes and send to GCS in `Raylet::RegisterGcs`. I think we can also delay initializing `gcs_client` until `RegisterGcs`, or make `ClientTableData` as a parameter of `Connect`.  E.g., 

```
/// Connect to the GCS. This method is used by non-raylet GcsClient. 
Status Connect();

/// Connect to the GCS and register ourself as a new node of the cluster.
/// This method is used by raylet.
Status ConnectAndRegisterNode(ClientTableData data);
```

By doing this, we can get rid of `ClientInfo`. And it's very clear that `ClientTableData` is only used by raylet.

Btw, I'd like to rename `ClientTableData` to `NodeInfo`. Because the name is very ambiguous.

Reply:
> Some states are still unavailable here. E..g, `node_manager_port` is 0 here, and will be set when the node manager grpc server starts (in the constructor of node manger).
> That's why we get these attributes and send to GCS in `Raylet::RegisterGcs`. I think we can also delay initializing `gcs_client` until `RegisterGcs`, or make `ClientTableData` as a parameter of `Connect`. E.g.,
> 
> ```
> /// Connect to the GCS. This method is used by non-raylet GcsClient. 
> Status Connect();
> 
> /// Connect to the GCS and register ourself as a new node of the cluster.
> /// This method is used by raylet.
> Status ConnectAndRegisterNode(ClientTableData data);
> ```
> 
> By doing this, we can get rid of `ClientInfo`. And it's very clear that `ClientTableData` is only used by raylet.
> 
> Btw, I'd like to rename `ClientTableData` to `NodeInfo`. Because the name is very ambiguous.

Got it. 

I think use ClientInfo is better than different Connect function. The interface should be as consistent as possible.

Or we can do it in:NodeStateAccessor, WorkerStateAccessor....

Reply:
> Some states are still unavailable here. E..g, `node_manager_port` is 0 here, and will be set when the node manager grpc server starts (in the constructor of node manger).
> That's why we get these attributes and send to GCS in `Raylet::RegisterGcs`. I think we can also delay initializing `gcs_client` until `RegisterGcs`, or make `ClientTableData` as a parameter of `Connect`. E.g.,
> 
> ```
> /// Connect to the GCS. This method is used by non-raylet GcsClient. 
> Status Connect();
> 
> /// Connect to the GCS and register ourself as a new node of the cluster.
> /// This method is used by raylet.
> Status ConnectAndRegisterNode(ClientTableData data);
> ```
> 
> By doing this, we can get rid of `ClientInfo`. And it's very clear that `ClientTableData` is only used by raylet.
> 
> Btw, I'd like to rename `ClientTableData` to `NodeInfo`. Because the name is very ambiguous.

I will make some change: 

Since in `main` we can't get enough information to register raylet to GCS, we will do `RegisterGcs` inside of raylet.

It's better use `NodeStateAccessor::Register` to do the node register job. And WorkerStateAccessor::Register to do de worker register job. Let's keep `GcsClient::Connect` as simple as possible(only do the job to connect to GCS), and make the interface clear.

And in this way, i can remove ClientInfo.